src/Pos/Generator/BlockEvent.hs

```diff
@@ -155,360 +339,49 @@ newtype Chance = Chance {getChance :: Rational}
-        toZeroBased i =
-            getBlockIndex i - getBlockIndex blockIndexStart
-        getBlock :: BlockIndex -> BlundDefault
-        getBlock i =
-            -- Indexing is safe here because by construction we have enough
-            -- blocks for the lookup to succeed.
-            getOldestFirst blocks !! toZeroBased i
-    return $ fmap getBlock <$> preBlockEvents
-
-data GenBlockEventState
-    = GbesStartingState
-    | GbesBlockchain (OldestFirst NE BlockIndex)
-    | GbesExpectedFailure
-
--- | A version of 'genBlockEvents' that generates event with block indices
--- instead of actual blocks.
-genBlockEvents' ::
-       (Monad m, RandomGen g)
-    => BlockEventGenParams
-    -> RandT g m [BlockEvent' BlockIndex]
-genBlockEvents' begp =
-    flip evalStateT GbesStartingState $ do
-        events <- replicateWhileM
-            (fromIntegral $ begp ^. begpBlockEventCount)
-            (genBlockEvent begp)
-        finalEvent <- get <&> \case
-            GbesStartingState -> []
-            GbesExpectedFailure -> []
-            GbesBlockchain blockchain ->
-                -- Rollback the entire blockchain.
-                [BlkEvRollback $ BlockEventRollback
-                    { _berInput    = toNewestFirst blockchain
-                    , _berOutValid = BlockRollbackSuccess
-                    }]
-        return $ events ++ finalEvent
-
-replicateWhileM :: Monad m => Int -> m (Maybe a) -> m [a]
-replicateWhileM n m = go n
-  where
-    go k | k < 1 = return []
-         | otherwise =
-             m >>= \case
-                 Nothing -> return []
-                 Just a  -> (a:) <$> go (k-1)
-
-data RollbackFailureType
-    = RftRollbackExcess
-    | RftRollbackDrop
-    deriving (Enum, Bounded)
-
-instance Random RollbackFailureType where
-    random = randomR (minBound, maxBound)
-    randomR (lo,hi) = runRand $ uniform [lo..hi]
-
-data ApplyFailureType
-    = AftApplyBad
-    | AftApplyNonCont
-    deriving (Enum, Bounded)
-
-instance Random ApplyFailureType where
-    random = randomR (minBound, maxBound)
-    randomR (lo,hi) = runRand $ uniform [lo..hi]
-
-newtype IsFailure = IsFailure Bool
-newtype IsRollback = IsRollback Bool
-
-genBlockEvent ::
-       (Monad m, RandomGen g)
-    => BlockEventGenParams
-    -> StateT GenBlockEventState (RandT g m) (Maybe (BlockEvent' BlockIndex))
-genBlockEvent begp = do
-    gbes <- get
-    failure <- lift $ IsFailure <$> byChance (begp ^. begpFailureChance)
-    rollback <- lift $ IsRollback <$> byChance (begp ^. begpRollbackChance)
-    (mBlockEvent, gbes') <- case gbes of
-        GbesExpectedFailure ->
-            return (Nothing, gbes)
-        GbesStartingState ->
-            genBlockStartingState failure rollback
-        GbesBlockchain blockchain ->
-            genBlockInBlockchain blockchain failure rollback
-    put gbes' $> mBlockEvent
+newtype CheckCount = CheckCount Word
+    deriving (Eq, Ord, Show, Num)
+
+blkEvTip :: BlockEvent -> Maybe HeaderHash
```

Comment:
Add a comment that this function returns the tip which is expected after the event is applied. Because it can also return the tip which is expected before the event.